BCP/Session Lifecycle Management.md

```diff
@@ -0,0 +1,147 @@
+|8|wallet update session|return `wallet_createSession` w/ new sessionId and no scopes **next time called** OR `wallet_sessionChanged` w/existing sessionId|`wallet_sessionChanged` w/o sessionId|
+
+## Lifecycle diagrams
+
+### Visualizing the lifecycle of a session without a `sessionId`
+
+```mermaid
+sequenceDiagram
+  rect rgb(255, 255, 255)
+  actor Caller
+  participant Wallet
+  participant WalletDataStore as Data Store (Wallet)
+  rect rgb(255, 245, 245)
+  note right of Caller: Session Creation
+  Caller->>Wallet: wallet_createSession
+  Wallet->>WalletDataStore: Persist session data
+  Wallet-->>Caller: {"sessionScopes": {...}}
+  end
+  rect rgb(245, 255, 245)
+  note right of Caller: Update Session
+  Caller->>Wallet: wallet_createSession (update auth)
+  Wallet->>WalletDataStore: Update session data
+  Wallet-->>Caller: {"sessionScopes": {updatedScopes...}}
+  end
+  rect rgb(245, 245, 255)
+  note right of Caller: Connection Interrupted w/ Wallet Side Session Modification
+  Caller-->Wallet: Connection Interrupted
+  Wallet->>WalletDataStore: User initiated session change
+  Wallet-->Caller: wallet_sessionChanged (attempt fails)
+  Caller-->Wallet: Connection Re-established
+  end
+  rect rgb(255, 245, 255)
+  note right of Caller: Get Session
+  Caller->>Wallet: wallet_getSession
+  Wallet-->>Caller: {"sessionScopes": {...}}
+  end
+  rect rgb(255, 245, 215)
+  note right of Caller: Revoke Session
+  Caller->>Wallet: wallet_revokeSession
+  Wallet->>WalletDataStore: Update session data
+  Wallet-->>Caller: {"result": "true"}
+  end
+  end
+```
+
+### Visualizing the lifecycle of a session **with** a `sessionId`
+
+```mermaid
+sequenceDiagram
+  rect rgb(255, 255, 255)
+  participant CallerDataStore as Data Store (Caller)
+  actor Caller
+  participant Wallet
+  participant WalletDataStore as Data Store (Wallet)
+  rect rgb(255, 245, 245)
+  note right of Caller: Session Creation
+  Caller->>Wallet: wallet_createSession
+  Wallet->>WalletDataStore: Persist session data
+  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {...}}
+  Caller->>CallerDataStore: Persist session data
+  end
+  rect rgb(245, 255, 245)
+  note right of Caller: Update Session
+  Caller->>Wallet: wallet_createSession (sessionId: 0xdeadbeef, {updatedScopes...})
+  Wallet->>WalletDataStore: Update session data
+  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {(updated)sessionScopes...}}
+  Caller->>CallerDataStore: Persist session data
+  end
+  rect rgb(245, 245, 255)
+  note right of Caller: User Initiated Session Change
+  Wallet->>WalletDataStore: User initiated session change
+  Wallet->>Caller: wallet_sessionChanged (sessionId: 0xdeadbeef)
+  Caller->>CallerDataStore: Update session data
+  end
+  rect rgb(255, 245, 255)
+  note right of Caller: Revoke Session
+  Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)
```

Comment:
I suppose this should be 
```suggestion
  Caller->>Wallet: wallet_createSession (no sessionId, no scopes)
```
@bumblefudge its not clear what a full revoke response should be?

Reply:
Actually both work-- if wallet SUPPORTS `wallet_revokeSession` (which the caller would know if it was authorized to call it in the scopeObject !), they could call it with the optional sessionId, or if the dapp doesn't know for a fact that i CAN explicitly use that method, any `wallet_createSession` with a new sessionId is expected to kill the previous sessionId-having session.  

(if I remember correctly, we didn't make it a MUST on the wallet side to kill multiple sessionId`s per dapp because it out would be hard to test/guarantee/prove.)

Reply:
```suggestion
  Caller->>Wallet: wallet_createSession (sessionId: 0xnewbeef, no scopes)
  Wallet->>WalletDataStore: Create new, empty session 0xnewbeef ; clear all older sessions w/same dapp
  Wallet-->>Caller: {"result": "true"} (session is revoked)
  Caller->>CallerDataStore: Clear session data
  end
  rect rgb(255, 245, 255)
  note right of Caller: Revoke Session
  Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)
```

Reply:
> if wallet SUPPORTS wallet_revokeSession (which the caller would know if it was authorized to call it in the scopeObject !)

Wait, do we want this to be a permissioned method? Seems like that these lifecycle methods should be non-permissioned?

Reply:
Oh, I guess here we're in the overlap of feature discovery overlapping with feature permissioning! What I meant was, as written, getting a connection **without** sessionId back from a wallet means a dapp can **assume** support for `wallet_revokeSession`, but getting a connection **with** sessionId back means... they don't know if they have to do it the legacy way or if they also have the option to do it the explicit way.  so a wallet could explicitly authorize that in `sessionScopes` (even if not requested!), but otherwise, how will the dapp know?

Reply:
But isn't `wallet_revokeSession` able to accept both with or without `sessionId`? If they get back the `sessionId`, that implicitly tells the dapp that the wallet is tracking that session with that id, so they should be calling it with that session id no? And then they can call it without `sessionId` if they don't get anything back?